src/tool_writeout_json.c

```diff
@@ -0,0 +1,201 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 1998 - 2020, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+#include "tool_setup.h"
+
+
+#ifdef HAVE_LOCALE_H
+#  include <locale.h>
+#endif
+
+
+#define ENABLE_CURLX_PRINTF
+
+
+/* use our own printf() functions */
+#include "curlx.h"
+#include "tool_cfgable.h"
+#include "tool_writeout_json.h"
+
+
+typedef enum {
+  JSON_NONE,
+  JSON_STRING,
+  JSON_LONG,
+  JSON_DOUBLE,
+  JSON_VERSION,
+  JSON_FILENAME
+} jsontype;
+
+struct mapping {
+  const char *key;
+  jsontype type;
+  CURLINFO cinfo;
+};
+
+static const struct mapping mappings[]={
+  {"url_effective", JSON_STRING, CURLINFO_EFFECTIVE_URL},
+  {"http_code", JSON_LONG, CURLINFO_RESPONSE_CODE},
+  {"response_code", JSON_LONG, CURLINFO_RESPONSE_CODE},
+  {"http_connect", JSON_LONG, CURLINFO_HTTP_CONNECTCODE},
+  {"time_total", JSON_DOUBLE, CURLINFO_TOTAL_TIME},
+  {"time_namelookup", JSON_DOUBLE, CURLINFO_NAMELOOKUP_TIME},
+  {"time_connect", JSON_DOUBLE, CURLINFO_CONNECT_TIME},
+  {"time_appconnect", JSON_DOUBLE, CURLINFO_APPCONNECT_TIME},
+  {"time_pretransfer", JSON_DOUBLE, CURLINFO_PRETRANSFER_TIME},
+  {"time_starttransfer", JSON_DOUBLE, CURLINFO_STARTTRANSFER_TIME},
+  {"size_header", JSON_LONG, CURLINFO_HEADER_SIZE},
+  {"size_request", JSON_LONG, CURLINFO_REQUEST_SIZE},
+  {"size_download", JSON_DOUBLE, CURLINFO_SIZE_DOWNLOAD},
```

Comment:
I think you should avoid the sizes as double and use the ones that return `curl_off_t` instead, as they're fine and clean 64 bit values. See the info options ending with `_T`.

Reply:
i oriented that to the original writeout code:

```
  case VAR_SIZE_DOWNLOAD:
  if(CURLE_OK ==
                   curl_easy_getinfo(curl, CURLINFO_SIZE_DOWNLOAD,
                                     &doubleinfo))
                  fprintf(stream, "%.0f", doubleinfo);
                break;
```

i was wondering in the first place how half-of-a-byte could end up there. for me, `size_download` is a long, but the other code indicates a double.

Reply:
ah, you meant to use CURLINFO_SIZE_DOWNLOAD_T instead of CURLINFO_SIZE_DOWNLOAD …

Reply:
yes exactly! There are `_T` alternatives for a bunch of the info variables. The double version ones are the original ones but aren't really the ones to prefer these days...

Reply:
works.